### DISCUSSION TITLE
FastAPI router decorator is skipping other decorators in case of 422 errors

### DISCUSSION THREAD

**Discussion by akshay4kumar** [fastapi/fastapi#10702] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import logging
from fastapi import FastAPI
import functools

app = FastAPI()

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)
logger.addHandler(logging.StreamHandler())

def authorization(func):
    @functools.wraps(func)
    async def wrapped_func(*args, **kwargs):
        logger.info("Do authorization and raise 403 if required")
        # authorization logic
        return await func(*args, **kwargs)
    return wrapped_func


@app.get("/app")
@authorization
async def get_app(app_name: str):
    return {"app_name": app_name}
```

### Description

In the above code when the endpoint GET /app is invoked without the compulsory query parameter app_name I'm getting a 422 Unprocessable Entity error, but I want the authorization to happen first and then the validation of parameters, due to security reasons.

I know the right way to do the authorization is via middleware of dependencies, but I want to know the reason for this behaviour.

### Operating System

macOS

### Operating System Details

ventura 13.0

### FastAPI Version

">=0.103.2,<1.0.0"

### Pydantic Version

version = "^1.10.2"

### Python Version

3.8.6

**Comment by jgould22** [comment]
>  but I want to know the reason for this behaviour

You have wrapped your decorator in the router decorator and the router decorator is what handles the validation 

https://github.com/tiangolo/fastapi/blob/master/fastapi/routing.py#L1289

**Reply by akshay4kumar** [reply]
okay so that means the router.get decorator is getting executed first that is why we are just ending up with the validation response.

Now a follow up question from here, 

when i flip the order of decorator in the above code

```
import logging
from fastapi import FastAPI
import functools

app = FastAPI()

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)
logger.addHandler(logging.StreamHandler())

def authorization(func):
    @functools.wraps(func)
    async def wrapped_func(*args, **kwargs):
        logger.info("Do authorization and raise 403 if required")
        # authorization logic
        return await func(*args, **kwargs)
    return wrapped_func


@authorization
@app.get("/app")
async def get_app(app_name: str):
    return {"app_name": app_name}
```

Now I'm noticing that in either of the case (if I pass the compulsory query param or not) my authz decorator is getting skipped, it is not getting executed.

Could please explain the reason for this behaviour, and if so is there any possible way I could implement this functionality through decorators.

**Reply by akshay4kumar** [reply]
@jgould22 Any comments on the above scenario.

**Comment by YuriiMotov** [comment, marked as answer]
Router decorator just adds the route to the list of routes.
When the client makes a request to the endpoint, FastAPI will find the route that matches requested path, then it will solve dependencies, and then execute your endpoint function.
The logic you added by `authorization` decorator will be executed right before the endpoint function, but it will happen after resolving dependencies.
You can not change this order..
